### DISCUSSION TITLE
所有與 pj1 課程相關的問題, 都必須先自行嘗試找答案, 若仍無法解決, 可在此發問.

### DISCUSSION THREAD

**Discussion by mdecycu** [mdecd2023/football-apj1#3] [Q&A]
提問任何問題之前, 請先看一下:

https://www.betterup.com/blog/how-to-ask-good-questions

**Comment by 41023224** [comment]
老師，請問我們之前遇到acp問題的原因是沒鑰匙，但是目前新的鑰匙已經辦好並且已設定完成，但還是無法acp，請問如何解決。

**Reply by scrum-1** [reply]
acp.bat 包含 
git add . (表示要新增所有改版內容)
git commit -m "提交訊息" (若使用沒有設定 home_ipv6/.gitconfig 中的 user.name 與 user.email, 將無法順利完成提交)
git push (這裡若採 ssh, 則其 session 包含連線到 github.com, 有沒有 key pair 對應, 該不該使用 proxy 等三項考量)

**Comment by scrum-2** [comment]
請先弄清楚 acp.bat 的指令內容, 然後再了解 ssh 協定 push 流程所需的條件, 一一檢查後, 就會知道問題出在哪.

**Comment by 41023247** [comment]
組員fork之後，要如何進到動態進行改版

**Reply by 41023247** [reply]
該問題已解決

**Reply by scrum-2** [reply]
就是要在近端 localhost 以 cms 執行動態網站改版, 或放到 stud2 主機中進行改版

**Comment by 41023224** [comment]
老師，請問假如組員不小心加入道組長的那個群組如何處理？有辦法退出嗎？

**Reply by RWE1116** [reply]
leave team

**Reply by mdecycu** [reply]
但是 ruoqinz0421 還是必須付出代價?

**Comment by RWE1116** [comment]
求球框與球之尺寸?

**Reply by scrum-2** [reply]
可以自行定義

**Reply by RWE1116** [reply]
好ㄉ

**Comment by Tzm0306** [comment]
![22](https://user-images.githubusercontent.com/91449945/227153035-d715b378-5758-487b-b2ad-e4c14d421e01.png)

請問程式碼顯示圖示之問題該如何除錯

**Reply by mdecycu** [reply]
可以將 ttt 或程式碼放上來, 應該是物件名稱打錯

**Reply by Tzm0306** [reply]
好的沒問題
[BubbleRob - 複製.zip](https://github.com/mdecd2023/football-apj1/files/11049314/BubbleRob.-.zip)

![33](https://user-images.githubusercontent.com/91449945/227165033-a4c2d0f2-9d2b-4619-893b-de9ac3f65618.png)

![11](https://user-images.githubusercontent.com/91449945/227165149-0ec5faa0-ab88-43c6-a677-8660f1496882.png)

**Reply by scrum-1** [reply]
lua 程式第 17 行物件名稱為 graph, 但樹狀架構中卻是 Graph
第 21 行 ui 介面中的 on-change 誤打為 onchange
修正上述兩個錯誤後, 程式可以執行, 但物件設定仍有其他錯誤.

**Reply by Tzm0306** [reply]
ok謝謝老師

**Comment by junpig10** [comment]
![1](https://user-images.githubusercontent.com/91455505/227163061-61a8f0d7-814d-467a-80a6-764d5f973c3e.png)

請問傳感器模擬時無法偵測該如何解決?

**Reply by scrum-1** [reply]
sensingNose 所回傳與障礙物的距離必須在控制程式中, 透過判斷式控制輪軸才能避障.

**Comment by jamie920130** [comment]
![image](https://user-images.githubusercontent.com/91661783/227818365-ceecec0d-1435-48ef-bdde-a0fc8b766652.png)
老師，我的轉靜態沒權限

**Reply by scrum-1** [reply]
先直接到網頁根目錄看看, 可能現在的網頁狀態是尚未登入.

**Comment by 41023105** [comment]
![圖片](https://user-images.githubusercontent.com/92246962/227824021-9e6d0511-b316-456e-9fed-9c96fd06491e.png)
老師，我們的物件無法移動

**Reply by scrum-1** [reply]
Geometric plugin 在正常的 CoppeliaSim 會自動載入, 若沒有載入, 可能表示該 CoppeliaSim 啟動程式有誤, 試著重新下載或者到其他正常的電腦上操作.

**Comment by yonqui0411** [comment]
在ACP的時候小黑框出現尚未上傳的影片，導致更新完動態內容，但acp錯誤

**Comment by 41023226** [comment]
將檔案下載下來後測試simpleTest_19997.py 後出現
![圖片](https://user-images.githubusercontent.com/91449738/228752836-b4406e10-bd45-441c-b0ed-184a8cd8bde8.png)

**Reply by Martin3130** [reply]
試試要到coppeliasim資料夾裡>system>uesr.txt>第172.173列改為defaultPython =   Y:/Python311/python.exe
executeUnsafe = true

**Reply by 41023226** [reply]
![圖片](https://user-images.githubusercontent.com/91449738/228772625-7e2549fc-d48c-4fee-a8d2-dea49ba3affe.png)
已改 但好像不是這個問題

**Reply by scrum-1** [reply]
上面修改 usrset.txt 的目的是要在場景中執行 Python child script, 而這裡的問題是要能夠執行 Legacy Remote API Python, 答案應該就在
https://mde.tw/cd2023/content/pj1.html 看看是否所有條件都已經符合

**Comment by CHIA-ROU** [comment]
在使用 TeXworks 轉成 Pdf 檔時出現因缺少 Moe Standard Kai 和 Nimbus Roman，而無法生成 Pdf
![image](https://user-images.githubusercontent.com/91234428/230810163-128dbfd5-d006-4dc4-9d31-07c0f68ce2c8.png)

**Reply by CHIA-ROU** [reply]
後來透過下載 https://www.moedict.tw/fonts/truetype/moe/ 中 MoeStandardKai.ttf 的檔案放入 MiKTeX\miktex\log 並安裝就可解決
但是下載 https://www.1001fonts.com/nimbus-roman-no9-l-font.html 中的檔案並安裝卻仍出現以下訊息
![image](https://user-images.githubusercontent.com/91234428/230814608-4f4c108d-962b-4a79-a166-e30702085fa8.png)

**Comment by jamie920130** [comment]
老師，我在筆電近態acp玩，我的倉儲仍未顯現有推上去
![mdecd2023_2a-pj1ag10_ 2a-pj1ag10 created by GitHub Classroom - Google Chrome 2023_4_10 下午 05_10_01](https://user-images.githubusercontent.com/91661783/230872587-4f85239c-e078-4074-bae5-a1118df0de05.png)
![2a-pj1ag10 — Mozilla Firefox 2023_4_10 下午 05_08_21](https://user-images.githubusercontent.com/91661783/230872606-c3778f41-7994-491e-91da-ea3eee137c62.png)

**Reply by RWE1116** [reply]
或許能看看黑色視窗有沒有上傳成功那邊比較準確，網頁上不管成功與否都會顯示這個畫面

**Comment by junpig10** [comment]
老師，acp之後沒有推成功，不知道問題在哪
![8](https://user-images.githubusercontent.com/91455505/232750347-5454f5cd-04b6-4697-bca5-58f65849b89a.jpg)

**Comment by nfu40723217** [comment]
![image](https://user-images.githubusercontent.com/126143777/233068320-113b8284-8705-4f1a-8ea8-108a6e87a553.png)
我嘗試在家裡使用CoppeliaSim去連接我自己的主機，但無論如何都無法解決系統找不到.py檔案路徑的問題，但在學校這個問題就得以解決，是不是python中出現了甚麼問題導致無法找到?

(已經自行使用chatgpt解決，下文文長注意)

**Comment by nfu40723217** [comment]
承接上文，我使用了chatgpt和三個小時的奮戰下，終於讓自己能夠在宿舍也能使用可攜系統進行bubbleRob的鍵盤操作了。
以下是解決的一半過程對話:(Q代指我、A代指chatgpt)

Q:我在Scite上遇到了"系統找不到指定的檔案。" 但我不清楚是甚麼原因，能列舉可能的原因嗎?

A:"系統找不到指定的檔案" 是一個錯誤訊息，通常代表了一個檔案或路徑並不存在或無法被找到。以下是可能導致這個錯誤的原因：

1.檔案或路徑不存在：這是最常見的原因，檔案或路徑可能已被刪除、移動或更改名稱。
2.權限不足：如果您不擁有足夠的權限來訪問該檔案或路徑，系統就會報告此錯誤。
3.檔案或路徑正在使用中：如果檔案或路徑正在被其他程式使用，例如正在被其他程式打開，您就無法訪問它。
4.磁碟問題：如果您的硬碟遇到了問題，例如壞道或損壞，也可能會導致系統找不到指定的檔案。
5.程式錯誤：有時候程式本身也可能導致這個錯誤，可能是因為程式中存在錯誤或不正確的路徑設置。

建議您檢查檔案路徑是否正確，確認您有足夠的權限來訪問檔案，並關閉其他正在使用檔案的程式。如果問題仍然存在，您可以嘗試重新啟動您的電腦，或檢查您的硬碟是否存在問題。

Q:那麼在Scite的輸出欄位中出現"pyw -u 文件"代表著甚麼意思?

A:在 Scite 的輸出欄位中出現 "pyw -u 文件" 代表著正在執行一個 Python 程式，其中：

"pyw" 是 Python 的執行檔案名稱，通常用於執行 GUI 應用程式。
"-u" 代表著執行 Python 程式時使用的參數，這裡 "-u" 是代表使用 unbuffered I/O（無緩衝輸入輸出）。
"文件" 是指正在執行的 Python 程式的檔案名稱。
因此，"pyw -u 文件" 意味著 Scite 正在使用 Python 執行檔案執行指定的 Python 程式檔案，並使用 unbuffered I/O 參數。

Q:那麼如果Scite執行了這個檔案中的代碼，並出現找不到指定檔案，你認為會是哪種問題?

A:如果在 Scite 中執行 Python 程式時出現 "系統找不到指定的檔案" 的錯誤訊息，通常代表著程式中的某個檔案無法被找到，可能是以下原因之一：

1.路徑不正確：程式中指定的路徑可能是不正確的，例如路徑名稱有誤、縮寫或空格使用不當等。
2.檔案不存在：程式中指定的檔案可能不存在，例如該檔案被刪除或被移動到了其他地方。
3.權限不足：程式無法訪問該檔案，可能是因為權限不足或該檔案被鎖定而無法被訪問。
4.檔案正在被其他程式使用：可能是因為該檔案正在被其他程式使用而無法被訪問。

如果程式中指定的檔案名稱和路徑都是正確的，可以檢查一下該檔案是否存在，並且您是否擁有足夠的權限來訪問該檔案。如果該檔案存在，請確保該檔案沒有被其他程式鎖定。另外，您也可以嘗試使用絕對路徑或相對路徑指定檔案的位置，以確保程式可以正確找到該檔案。

//到這裡，看來我白問了不少問題。

Q:我沒有移動檔案，單純只是在Scite的上輸入python代碼 print("hi") ，然後儲存在Scite的根目錄上，卻還是發生上述問題，該如何解決?

A:如果您在 Scite 的根目錄中建立一個 Python 檔案，並在該檔案中加入 print("hi") 代碼，然後在 Scite 中運行該檔案時出現了 "系統找不到指定的檔案" 的錯誤，這可能是因為 Scite 沒有正確設置當前工作目錄。

請確保您正在使用的 Scite 版本支援當前工作目錄設置，並且您已經正確地設置了當前工作目錄。您可以在 Scite 中通過點擊 "Options" 菜單中的 "Open User Options File" 來編輯 Scite 的用戶設置文件，然後添加以下行：
`cwd=$(FilePath)`
這個設置將當前工作目錄設置為當前打開的檔案所在的目錄。這樣，當您運行 Python 程式時，Python 將會在當前工作目錄中搜索檔案，而不是在 Scite 的根目錄中搜索檔案，從而避免 "系統找不到指定的檔案" 的錯誤。

Q:我已經嘗試加入了以上代碼
目前我的環境如下(以基本目錄為Y硬碟):
Y:\wscite 為Scite目錄
Y:\Python310 為python3.10.0目錄
而python代碼 print("hi") 儲存名稱為 test.py 並儲存在scite目錄下
但依舊出現同樣的問題，該如何解決?

A:如果您已經在 Scite 的用戶設置文件中設置了當前工作目錄，但仍然出現 "系統找不到指定的檔案" 的錯誤，您可以試試以下解決方案：
1.確保 Python 已經安裝並設置好環境變數：請確保 Python 已經成功安裝在您的電腦上，並且已經設置好環境變數，以便您可以在命令提示符號或 Scite 中運行 Python 程式。
2.檢查 Scite 設置：請確保 Scite 的設置正確，例如編譯器路徑、Python 版本等。
3.檢查 Python 程式碼：請確保您的 Python 程式碼沒有錯誤，例如語法錯誤或拼寫錯誤等。
4.嘗試使用絕對路徑：請嘗試使用絕對路徑指定 Python 程式碼所在的位置，以確保 Scite 可以正確找到該檔案。例如，在 Scite 中使用以下代碼來執行您的 Python 程式碼：
`import os
os.system("Y:/wscite/Python310/python.exe Y:/wscite/test.py")`
這個代碼將 Python 程式碼 test.py 所在的絕對路徑傳遞給 Python 解釋器，從而避免了 "系統找不到指定的檔案" 的錯誤。
如果以上方法仍然無法解決問題，您可以嘗試重新安裝 Scite 和 Python，或者嘗試使用其他的程式碼編輯器來編輯和運行 Python 程式碼。

Q:我發現它輸入此程式碼後跳亂碼了......

A:如果您在 Scite 中輸入 Python 程式碼後出現亂碼，可能是因為 Scite 的編碼格式與您的 Python 程式碼的編碼格式不一致所導致的。
您可以嘗試以下方法解決問題：
設置 Scite 的編碼格式：打開 Scite，從菜單欄選擇「選項」 >「編輯器設置」 >「默認編碼」，選擇與您的 Python 程式碼相同的編碼格式。
轉換 Python 程式碼的編碼格式：使用一些編輯器工具（例如 Notepad++）將您的 Python 程式碼轉換為 Scite 使用的編碼格式，以確保 Scite 正確顯示您的程式碼。
在程式碼中添加編碼宣告：在 Python 程式碼的開頭添加一行編碼宣告，例如：
`# -*- coding: utf-8 -*-`
這個宣告將告訴 Python 程式碼的解釋器使用 UTF-8 編碼格式來解讀程式碼。
希望以上方法能夠幫助您解決問題。

Q:在運行 import os
os.system("Y:/wscite/Python310/python.exe Y:/wscite/test.py")
程式碼後，問題依舊是找不到指定的檔案
要不然試著一個個排除吧

在那之後就是一大段的排除錯誤。

當我進行到
A:如果在打開 Open User Options 後看到空白的介面，這代表你的 Scite 還沒有任何設置選項。你可以手動添加設置選項，方法如下：

創建一個空白的文本檔案。
將以下內容複製到文件中：

`//This is an example SciTE properties file with a couple of settings`
`//that show how to configure SciTE to work with Python.`
`//Define the location of the Python executable.`
`command.go.*.py=python3 -u $(FileNameExt)`
`//Enable the Python lexer.`
`lexer.$(file.patterns.py)=python`
`//Set the encoding for Python source code files.`
`code.page=65001`
將文件保存到 Scite 的配置文件夾中。你可以通過選擇「選項」>「打開配置文件夾」找到配置文件夾。在文件名輸入框中輸入「SciTEUser.properties」，然後點擊「保存」。

重新啟動 Scite，你應該可以看到新添加的選項已經生效了。
注意：如果你的 Python 安裝路徑與預設路徑不同，你需要修改以上代碼中的 Python 執行路徑。

接著，我的SciTE終於開始運作了！

如果還想看以後的討論介紹，請到bg11進行觀看。
[https://mdecd2023.github.io/2b-pj1bg11/content/0420-1hr.html](url)
